File: solutions/typescript/resistor-color-duo/1/resistor-color-duo.ts
export function decodedValue(colors: string[]): number {
  return colors
      .slice(0, 2)
      .reverse()
      .reduce(
        (acc: number, color: string, index: number) => acc + COLORS.indexOf(color) * 10 ** index,
        0
      )
}

const COLORS = [
  "black",
  "brown",
  "red",
  "orange",
  "yellow",
  "green",
  "blue",
  "violet",
  "grey",
  "white",
]

File: solutions/prolog/resistor-color-duo/2/resistor_color_duo.pl
value(Colors, Value) :-
    [First, Second | _] = Colors,
    maplist(color_code, [First, Second], Digits),
    foldl(append_digit, Digits, 0, Value).

append_digit(Digit, Acc, Value) :- Value is Acc * 10 + Digit. 

color_code(Color, Code) :-
    colors(AllColors),
    nth0(Code, AllColors, Color).

colors(Colors) :-
    Colors = [
        "black", "brown", "red", "orange", "yellow", 
        "green", "blue", "violet", "grey", "white"
    ].
